### PR TITLE
decision in the matter of cloudmon/stackmon approach - review

### PR DESCRIPTION
**Pull request by piobig2871**

### CODE REVIEW

File: Standards/scs-0106-v1-evaluate-current-state-of-cloudmon.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+clear, human-readable test behavior definitions, facilitating participation from both developers and non-technical
+contributors. Considering the team's proficiency in Python, the language's simplicity and clarity, alignment with
+OpenStack's ecosystem, and the robust support from the Python community, it's evident that Python presents a superior
+choice for implementing Gherkin-based testing over Java. By harnessing Python's strengths, we can maximize efficiency,
+accelerate development, and ensure seamless integration with OpenStack, ultimately enhancing the effectiveness of our
+testing processes.
+
+## Decision
+
+By opting for Gherkin test scenarios with Python mapping API calls to OpenStack, we aim to address the
```
Comment (artificial-intelligence) [inline Standards/scs-0106-v1-evaluate-current-state-of-cloudmon.md:39]: we could link to Gherkin here, maybe, for people not familiar with it (myself included)?
I believe this should be an official source? https://cucumber.io/docs/gherkin/

Comment (piobig2871) [reply Standards/scs-0106-v1-evaluate-current-state-of-cloudmon.md:39]: @artificial-intelligence this is a very good idea. I have not thought about that, thanks for pointing it out. I can also write a sentence or two about behave framework (https://behave.readthedocs.io/en/latest/) that we are using and link the documentation there. Would that be helpful as well?

Comment (artificial-intelligence) [reply Standards/scs-0106-v1-evaluate-current-state-of-cloudmon.md:39]: Yes that would be nice, I'm only lightly familiar with the topic (I think I read some articles about it some years ago).

File: Standards/scs-0106-v1-evaluate-current-state-of-cloudmon.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+of cloud infrastructure is paramount for organizations. Effective monitoring and testing framework
+are essential tools in this endeavor, enabling proactive identification and resolution of issues before
+they impact critical operations. However, selecting the right monitoring solution can be a daunting task,
+particularly when faced with options that are overly complex and lack essential documentation and working
+scenarios.
+
+## Motivation
+
+One such solution, the [Cloudmon](https://stackmon.org/) project, presents challenges that may hinder its
+suitability for organizations seeking efficient and reliable monitoring capabilities. This introduction
```
Comment (bitkeks) [inline Standards/scs-0106-v1-evaluate-current-state-of-cloudmon.md:18]: Should we list the challenges? The tool itself does good work - for infrastructures of a bigger scale like OTC. We need to compare it to our use case to put the *challenges* into context

File: Standards/scs-0106-v1-evaluate-current-state-of-cloudmon.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+contributors. Considering the team's proficiency in Python, the language's simplicity and clarity, alignment with
+OpenStack's ecosystem, and the robust support from the Python community, it's evident that Python presents a superior
+choice for implementing Gherkin-based testing over Java. By harnessing Python's strengths, we can maximize efficiency,
+accelerate development, and ensure seamless integration with OpenStack, ultimately enhancing the effectiveness of our
+testing processes.
+
+## Decision
+
+By opting for [Gherkin](https://cucumber.io/docs/gherkin/) test scenarios with Python mapping API calls to OpenStack,
+we aim to address the complexities and shortcomings of the Cloudmon project while ensuring our monitoring and testing
```
Comment (bitkeks) [inline Standards/scs-0106-v1-evaluate-current-state-of-cloudmon.md:40]: > we aim to address the complexities and shortcomings of the Cloudmon project

See comment above - what do these complexities and shortcomings consist of?

File: Standards/scs-0106-v1-evaluate-current-state-of-cloudmon.md (hunk 2 above)
Comment (piobig2871) [reply Standards/scs-0106-v1-evaluate-current-state-of-cloudmon.md:18]: I will describe that

File: Standards/scs-0106-v1-evaluate-current-state-of-cloudmon.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+testing processes.
+
+## Challanges
+
+During our assessment of the Cloudmon/Stackmon project, we encountered significant challenges related to documentation,
+particularly regarding lack of examples for configuration setups and usage guidelines. The lack of comprehensive
+documentation impeded our understanding of the project and hindered effective utilization. We understood that some
+features implemented in the cloudmon/stackmon are not neccessary for the health monitor like running the tests from
+different virtual locations because every network in the project is responsible for different AZ and on the other hand
+with our tests approach we can just clone [shs-health-monitor](https://github.com/SovereignCloudStack/scs-health-monitor/tree/main)
```
Comment (matofeder) [inline Standards/scs-0106-v1-evaluate-current-state-of-cloudmon.md:44]:
```suggestion
with our tests approach we can just clone [scs-health-monitor](https://github.com/SovereignCloudStack/scs-health-monitor/tree/main)
```

File: Standards/scs-0106-v1-evaluate-current-state-of-cloudmon.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+are essential tools in this endeavor, enabling proactive identification and resolution of issues before
+they impact critical operations. However, selecting the right monitoring solution can be a daunting task,
+particularly when faced with options that are overly complex and lack essential documentation and working
+scenarios.
+
+## Motivation
+
+One such solution, the [Cloudmon](https://stackmon.org/) project, presents challenges that may hinder its
+suitability for organizations seeking efficient and reliable monitoring capabilities. This introduction
+outlines the reasons why our organization opts against utilizing the Cloudmon project and instead embraces
```
Comment (matofeder) [inline Standards/scs-0106-v1-evaluate-current-state-of-cloudmon.md:19]: this sentence sounds more like a decision and not like a motivation

Comment (piobig2871) [reply Standards/scs-0106-v1-evaluate-current-state-of-cloudmon.md:19]: let me rephrase that, thanks

**Review by matofeder — APPROVED** [review]

File: Standards/scs-0106-v1-evaluate-current-state-of-cloudmon.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+scenarios.
+
+## Motivation
+
+One such solution, the [Cloudmon](https://stackmon.org/) project, faces challenges that might limit its effectiveness for organizations looking
+for efficient and reliable monitoring capabilities. This introduction outlines the reasons why our organization opts
+against utilizing the Cloudmon project and instead embraces a more streamlined and effective approach involving Gherkin
+test scenarios and mapping Python API calls to interact with OpenStack resources. By addressing the complexities and
+shortcomings of the Cloudmon project, SCS organization aims to adopt a monitoring solution that not only meets but
+exceeds our requirements for simplicity, reliability, and ease of use.
```
Comment (gtema) [inline Standards/scs-0106-v1-evaluate-current-state-of-cloudmon.md:22]: Honestly this phrasing is more then subjective. After dozens of hours spend explaining the documentation, setup scenarios, testing capabilities, workshops this statement is very offensive. SCS organization entered the evaluation and asked for multiple workshops to understand the tooling and offered help in improving the documentation. There was absolutely no output.
What you offer as an alternative is another "one of thousands" of namely universal testing frameworks that were not designed to monitor cloud environment and are not really dealing with measuring low level performance (of the API itself).

File: Standards/scs-0106-v1-evaluate-current-state-of-cloudmon.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+## Design Considerations
+
+Our approach was to base on a technical concept. This document serves as a proposal, with the final decision
+subject to discussion with SCS team members. We propose a behavior-driven system based on solutions using Java framework
+Cucumber, utilizing the Gherkin domain-specific language for defining executable specifications. This approach ensures
```
Comment (gtema) [inline Standards/scs-0106-v1-evaluate-current-state-of-cloudmon.md:28]: So what you offer is clearly controversial to the objective you stated above: simplicity. Taking Java and a widely unknown framework with even less known DSL is going to harm simplicity and understand-ability of the solution. It brings enormous amount of efforts required by operators to learn those things and be able to extend it.

File: Standards/scs-0106-v1-evaluate-current-state-of-cloudmon.md [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+clear, human-readable test behavior definitions, facilitating participation from both developers and non-technical
+contributors. Considering the team's proficiency in Python, the language's simplicity and clarity, alignment with
+OpenStack's ecosystem, and the robust support from the Python community, it's evident that Python presents a superior
```
Comment (gtema) [inline Standards/scs-0106-v1-evaluate-current-state-of-cloudmon.md:31]: I do not get how you come from Java mentioned above to the python

File: Standards/scs-0106-v1-evaluate-current-state-of-cloudmon.md [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+choice for implementing Gherkin-based testing over Java. By harnessing Python's strengths, we can maximize efficiency,
+accelerate development, and ensure seamless integration with OpenStack, ultimately enhancing the effectiveness of our
+testing processes.
+
+## Challanges
+
+During our assessment of the Cloudmon/Stackmon project, we encountered significant challenges related to documentation,
+particularly regarding lack of examples for configuration setups and usage guidelines. The lack of comprehensive
```
Comment (gtema) [inline Standards/scs-0106-v1-evaluate-current-state-of-cloudmon.md:39]: This doesn't sound fair. You were offered dedicated session, pointed to the relevant documentation (and as mentioned above it was communicated towards the CloudMon team that taking those workshops you will help improving the documentation), pointed to detailed explanation of the configuration, pointed to the live production configuration. If production configuration is not considered as "lack of examples" I do not know what to say more. If production testing scenarios of a very big OpenStack based public cloud with much more services then vanila SCS comes with are not sufficient it shows me the evaluation was performed not very detailed.

File: Standards/scs-0106-v1-evaluate-current-state-of-cloudmon.md [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+documentation impeded our understanding of the project and hindered effective utilization. We understood that some
+features implemented in the cloudmon/stackmon are not neccessary for the health monitor like running the tests from
+different virtual locations because every network in the project is responsible for different AZ and on the other hand
+with our tests approach we can just clone [scs-health-monitor](https://github.com/SovereignCloudStack/scs-health-monitor/tree/main)
+to another physical location and run the same tests from physically different place on earth. Examples like that
+confirmed our belief that atomic approach save us a lot of effort, time and costs.
+
+## Decision
+
+By opting for [Gherkin](https://cucumber.io/docs/gherkin/) test scenarios with Python mapping API calls to OpenStack,
```
Comment (gtema) [inline Standards/scs-0106-v1-evaluate-current-state-of-cloudmon.md:49]: OpenTelekomCloud is using alternative framework (Robot Framework) for doing system acceptance testing. This comes with a huge effort (which you do not explicitly mention here, but is read between the lines) of developing this "python mapping API calls to OpenStack". This is a huge and complex work that can be avoided by relying on existing tooling for OpenStack. And btw, RobotFramework (at OTC) and behavior driven testing is great in testing, but is absolutely unsuitable for health and performance monitoring

**Review by gtema — CHANGES_REQUESTED** [review]
This is very subjective outcome and proposed alternative does look good to me while bringing enormous effort for the implementation.

File: Standards/scs-0106-v1-evaluate-current-state-of-cloudmon.md [hunk 11]
```diff
@@ -0,0 +1,53 @@
+---
+title: Evaluating the deployment of CloudMon in SCS infrastructure
+type: Decision Record
+---
+
+## Introduction
+
+In the fast-paced environment of modern cloud computing, ensuring the reliability and performance
+of cloud infrastructure is paramount for organizations. Effective monitoring and testing framework
+are essential tools in this endeavor, enabling proactive identification and resolution of issues before
```
Comment (artificial-intelligence) [inline Standards/scs-0106-v1-evaluate-current-state-of-cloudmon.md:10]: e.g. most monitoring doesn't really allow proactive resolution of issues.

Monitoring is just analysing live data after the fact. e.g. you can see an API is slow in monitoring, or you can see a disk filling up. But all the activities monitoring allows a DevOps Team to do are _reactive_. You notice a problem in monitoring, then you react.

A testing framework, on the other hand, _can_ of course be used to detect problems e.g. before deployment.

File: Standards/scs-0106-v1-evaluate-current-state-of-cloudmon.md [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+for efficient and reliable monitoring capabilities. This introduction outlines the reasons why our organization opts
+against utilizing the Cloudmon project and instead embraces a more streamlined and effective approach involving Gherkin
+test scenarios and mapping Python API calls to interact with OpenStack resources. By addressing the complexities and
+shortcomings of the Cloudmon project, SCS organization aims to adopt a monitoring solution that not only meets but
+exceeds our requirements for simplicity, reliability, and ease of use.
+
+## Design Considerations
+
+Our approach was to base on a technical concept. This document serves as a proposal, with the final decision
+subject to discussion with SCS team members. We propose a behavior-driven system based on solutions using Java framework
```
Comment (artificial-intelligence) [inline Standards/scs-0106-v1-evaluate-current-state-of-cloudmon.md:27]: There are many different testing frameworks and testing strategies, which are not really differentiated in this text. You propose a behavior-driven test concept using cucumber, but somehow the document fails to mention any up- or downsides of behavior driven testing or of this specific framework.

There is no reasoning I have read in this text why behavior driven testing is superior to the cloudmon approach.

And I don't doubt that there might be advantages to it, but those need to be clearly spelled out somewhere.

This text also talks about a "technical concept", where can I see that? It isn't linked here, or included. If it is the basis for the decision, it surely should be included?

I only know both technologies superficially, for the record, that is cloudmon and cucumber.

**Comment by artificial-intelligence** [comment]
As a general comment on how to approach system design decisions:

From this text, it is not clear to me, why _exactly_ is cucumber superior.

I don't think it's a matter of the implementation language.

You should be able to show concrete examples what cucumber actually does better than cloudmon, so you could convince me, as a reader.

I would also expect a proposal to use new technology to list the shortcomings of said technology as well, because in my experience every technological decision has, of course, up- and downsides.

Presenting only upsides is showing me, that we either are not aware of the downsides, which is dangerous, because it means we will learn of the downsides much later, and they might be even worse then the previous solution. In rare cases there are no downsides and you have a straight up better technical solution, which should be spelled out as well in the text, showing that at least it was thought about if there are downsides to the new solution.

If I don't see such things, it's a red flag to me, as it's shows a clear bias in the decision, because when we are honest, of course we all, as humans, like one technology more than the other for a variety of subjective reasons like prior familiarity, trendiness, marketing and last but not least, technical merit.

I think though it's important that we know about our own biases and at least make an attempt to counter them by an honest technical analysis of the up- and downsides of a given software.

So either not enough research was done, to be able to list up- and downsides of all solutions.
Or there is some bias, sweeping downsides under the rug.

Both are understandable things that happen to all of us, still I think if we want to make better decisions we should try to avoid that.

Thanks for reading, if you made it this far.

**Review by artificial-intelligence — CHANGES_REQUESTED** [review]

File: Standards/scs-0106-v1-evaluate-current-state-of-cloudmon.md (hunk 7 above)
Comment (gtema) [reply Standards/scs-0106-v1-evaluate-current-state-of-cloudmon.md:28]: main idea of stackmon is in using ansible as a test scenario that literally anybody is able to read/understand/replay locally/extend (endlessly)

**Comment by bitkeks** [comment]
What's the status here?

**Comment by bitkeks** [comment]
Ping @artificial-intelligence @gtema

**Comment by mbuechse** [comment]
Closing in favor of https://github.com/SovereignCloudStack/scs-health-monitor/pull/146
